exporters/dynamicconfig/confignotifier.go

```diff
@@ -55,45 +68,114 @@ type ConfigWatcher interface {
 }
 
 type ConfigNotifier struct {
+	ch             chan struct{}
 	checkFrequency time.Duration
+	clock          controllerTime.Clock
 	config         *MetricConfig
+	configHost     string
+	lock           sync.Mutex
 	subscribed     map[ConfigWatcher]bool
+	ticker         controllerTime.Ticker
+	wg             sync.WaitGroup
 }
 
-// Constructor for a ConfigNotifier, also starts it running
-func New(checkFrequency time.Duration, config *MetricConfig) *ConfigNotifier {
+// Constructor for a ConfigNotifier
+// Set configHost to "" if there is no remote config service and the config is not dynamic
+func New(checkFrequency time.Duration, defaultConfig *MetricConfig, configHost string) *ConfigNotifier {
+	config := defaultConfig
+	if configHost != "" {
+		config = readConfig(configHost)
+	}
+
 	configNotifier := &ConfigNotifier{
+		ch:             make(chan struct{}),
 		checkFrequency: checkFrequency,
+		clock:          controllerTime.RealClock{},
 		config:         config,
+		configHost:     configHost,
 		subscribed:     make(map[ConfigWatcher]bool),
 	}
 
-	go configNotifier.checkChanges()
-
 	return configNotifier
 }
 
-func (notifier *ConfigNotifier) Register(watcher ConfigWatcher) {
-	notifier.subscribed[watcher] = true
-	watcher.OnInitialConfig(notifier.config)
+// SetClock supports setting a mock clock for testing.  This must be
+// called before Start().
+func (n *ConfigNotifier) SetClock(clock controllerTime.Clock) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.clock = clock
 }
 
-func (notifier *ConfigNotifier) Unregister(watcher ConfigWatcher) {
-	delete(notifier.subscribed, watcher)
+func (n *ConfigNotifier) Start() {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	if n.configHost == "" {
+		return
+	}
+
+	if n.ticker != nil {
+		return
+	}
+
+	n.ticker = n.clock.Ticker(n.checkFrequency)
+	n.wg.Add(1)
+	go n.checkChanges(n.ch)
 }
 
-func (notifier *ConfigNotifier) checkChanges() {
-	for {
-		time.Sleep(notifier.checkFrequency)
+func (n *ConfigNotifier) Stop() {
+	n.lock.Lock()
+
+	if n.configHost == "" {
+		return
+	}
+
+	if n.ch == nil {
+		return
+	}
+	close(n.ch)
+	n.ch = nil
+
+	n.lock.Unlock()
 
-		newConfig := readConfig()
+	n.wg.Wait()
+	n.ticker.Stop()
+}
+
+func (n *ConfigNotifier) Register(watcher ConfigWatcher) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	n.subscribed[watcher] = true
+	watcher.OnInitialConfig(n.config)
```

Comment:
Let's don't hold the lock while calling "watcher.OnInitialConfig" because who knows how long the watcher is going to fiddle with the config.